webRTC 기능 추가

bhw1994 opened the pull request:
### **issue #2**

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/model/dao/UserDAO.java:
```diff
@@ -0,0 +1,90 @@
+                Map<String, Object> data = new HashMap<>();
+                data.put(KEY_NAME, name);
+                data.put(KEY_TOKEN, s);
+                data.put(KEY_TEL, "010-0000-0000");
```
이 부분도 마찬가지로 객체로 저장할 수 있으면 좋을 듯 합니다.
추가로 flatMap로 여러번 처리하는 것 보다는 리스트를 받는것도 괜찮아 보입니다.

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/api/firebase/FireStoreManager.java:
```diff
@@ -0,0 +1,58 @@
+        if(documentId==AUTO_CREATE_DOCUMENT_ID)
+            task=db.collection(collection).add(data);
+        else
+            task=db.collection(collection).document(documentId).set(data);
```
add는 무조건 문서가 추가되고 set이 해당 데이터가 이미 있으면 덮어 씌워주는거로 알고 있습니다.
나눠서 처리한 부분에 이유가 있나요?

bhw1994 replied:
add는 문서 ID를 자동으로 생성해 주는 기능으로 사용하였고 , document.set 은 문서 id를 직접 지정하는 용도로 사용한 것입니다.

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/api/firebase/FireStoreManager.java:
```diff
@@ -0,0 +1,58 @@
+
+
+        return Single.create(emitter -> task.addOnCompleteListener(task1 -> emitter.onSuccess(COMPLETE_STRING))
+                .addOnFailureListener(e -> emitter.onError(e)));
```
`Single`에서 지원하는 `fromCallable`로 반환형으로 값을 바로 넘겨주는 방법이 있습니다.
혹시 접목이 되지 않을까 싶어 남겨봅니다,.

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/api/firebase/FireStoreManager.java:
```diff
@@ -0,0 +1,58 @@
+
+
+
+}
```
불필요한 개행은 빼는게 좋아 보입니다,.

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/model/dao/UserDAO.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+
+    public static Single<List<User>> readAll(){
```
지난번에 말씀하신 것 처럼 비동기 안에서 비동기 처리가 불안하다면
`get()`이나 `collection()`을 리턴하여 `Reference`로 처리하는것도 나쁘지 않아 보입니다.

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/model/dao/UserDAO.java:
```diff
@@ -0,0 +1,90 @@
+
+public class UserDAO {
+    private static String TAG="BOYJ_UserInfoDAO";
+    public static String COLLECTION_NAME="User";
```
변수 명이 `User`를 가리키기에는 너무 일반적이라고 생각됩니다.

tbtzpdlql commented on app/src/main/java/com/webrtc/boyj/model/dao/UserDAO.java:
```diff
@@ -0,0 +1,90 @@
+import io.reactivex.functions.Function;
+
+public class UserDAO {
+    private static String TAG="BOYJ_UserInfoDAO";
```
`Logger`클래스가 `Util`에 있습니다.  그것을 활용하면 태그를 따로 만들 필요가 없어 보입니다.

tbtzpdlql commented on app/src/androidTest/java/com/webrtc/boyj/api/firebase/FireStoreManagerTest.java:
```diff
@@ -0,0 +1,47 @@
+        HashMap data=new HashMap<>();
+        data.put("testKey","TestValue");
+        FireStoreManager.createDocument("TestCollection","TestDocument",data).subscribe(result -> {
+            future.complete((String) result);
```
테스트는 `Future`로밖에 실행이 안되나요?

bhw1994 replied:
비동기 실행결과를 체크해야 했으므로 Future 를 사용했습니다.
대안이있으면 남겨주세요

tbtzpdlql commented on build.gradle:
```diff
@@ -12,7 +12,9 @@ buildscript {
-        
+
+        classpath 'com.google.gms:google-services:4.0.1'
+        //version 4.1.0 occured error;
```
제가 마지막에 `4.2.0`으로 잘 돌아갔었는데 `4.1.0`의 어떤부분이 문제가 생겼나요?

shooes270 commented on app/src/androidTest/java/com/webrtc/boyj/api/firebase/FireStoreManagerTest.java:
```diff
@@ -0,0 +1,47 @@
+
+public class FireStoreManagerTest {
+    @Test
+    public void getCollection() {
```
유닛테스트의 이름은 should로 시작하여 테스트를 잘 설명할 수 있어야합니다.
shouldEqualCollectionName 정도가 적당해보입니다.

shooes270 commented on app/src/androidTest/java/com/webrtc/boyj/api/firebase/FireStoreManagerTest.java:
```diff
@@ -0,0 +1,47 @@
+        try {
+            assertEquals(FireStoreManager.COMPLETE_STRING,future.get());
+        } catch (ExecutionException e) {
+            e.printStackTrace();
```
e.printStackTrace는 예외핸들링이 아닙니다.
테스트중 예측 가능한 (예상에 맞는 에러)와 예측 불가능한 에러의 처리방법이 다르지만
별도의 테스트 성공, 실패에 대한 구문을 추가하는것이 좋습니다

shooes270 commented on app/src/androidTest/java/com/webrtc/boyj/api/firebase/FireStoreManagerTest.java:
```diff
@@ -0,0 +1,47 @@
+        CompletableFuture<String> future=new CompletableFuture<>();
+        HashMap data=new HashMap<>();
+        data.put("testKey","TestValue");
+        FireStoreManager.createDocument("TestCollection","TestDocument",data).subscribe(result -> {
```
static 메소드의 테스트의 경우 테스트 순서에 따라 결과에 영향을 줄 수 있는데 이에대한 검증이 있었나요?

bhw1994 commented on app/build.gradle:
```diff
@@ -45,4 +66,12 @@ dependencies {
+    implementation 'com.android.support:multidex:1.0.3'
+
 }
+apply plugin: 'com.google.gms.google-services'
```
https://firebase.google.com/support/guides/google-android?hl=ko#update_your_gradle_configuration

공식 문서를 참고하여도 일반적으로 맨 하단에 추가하는것을 알수 있습니다

bhw1994 committed:
add webRTC feature.
    add webrtc dependency
    add retrofit , gson , okhttp dependency for get ice server list.
    add peerconnetion client.
    add ice server

    add camera , audio permission for webrtc
    add App class for global context

laco-dev commented on app/build.gradle:
```diff
@@ -45,4 +45,11 @@ dependencies {
     // RxJava
     implementation "io.reactivex.rxjava2:rxandroid:$rxandroid_version"
     implementation "io.reactivex.rxjava2:rxjava:$rxjava_version"
+    implementation "org.webrtc:google-webrtc:$webrtc_version"
```
해당 라이브러리 앞에 주석으로 설명 간단하게 추가 부탁드립니다.

버전을 변수명으로 사용하면 다른 경우에 대해서도 통일 해주세요!

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/data/IceServer.java:
```diff
@@ -0,0 +1,13 @@
+public class IceServer {
+    public String uri;
+    public String hostname;
+    public String password;
```
데이터는 private으로 하는게 어떨까요?

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/data/stun/IceServer.java:
```diff
@@ -0,0 +1,27 @@
+    @SerializedName("username")
+    @Expose
+    public String username;
+    @SerializedName("credential")
```
Json 필드와 클래스 필드가 동일하다면 `@SerializedName`은 필요 없습니다.

seokhyun-oh replied:
프로퍼티간의 구분이 힘드네요. 한 줄씩 구분을 해주면 보기 좋을거같습니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/data/stun/TurnServer.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface TurnServer {
+    @PUT("/_turn/TheBOYJ")
+    Call<TurnServerPojo> getIceCandidates(@Header("Authorization") String authkey);
```
`retrofit`은 RxJava와 호환이 됩니다. 
이를 통해 큐 형식이 아닌 옵저버블로 결과를 받아 처리할 수 있습니다.
해당 내용 알아보시고 적용하는게 좋아보입니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/data/stun/TurnServerPojo.java:
```diff
@@ -0,0 +1,30 @@
+    public String p;
+    @SerializedName("e")
+    @Expose
+    public Object e;
```
해당 POJO도 마찬가지로 private이 좋아보입니다.

seokhyun-oh replied:
POJO를 만족하는 클래스인건 알겠으나 클래스 명에 명시할 필요는 없어보입니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+
+    public PublishSubject<SessionDescription> sdpSubject=PublishSubject.create();
+    public PublishSubject<IceCandidate> iceCandidateSubject=PublishSubject.create();
+    public PublishSubject<MediaStream> remoteMediaStreamSubject=PublishSubject.create();
```
린트 적용 하게되면 = 부분에 대해 잡아주는것도 필요할 듯 합니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+                super.onCreateSuccess(sessionDescription);
+                localPeer.setLocalDescription(new CustomSdpObserver("localSetLocalDesc"), sessionDescription);
+                //SignalingClient.getInstance().emitSessionDescription(sessionDescription);
+                sdpSubject.onNext(sessionDescription);
```
offer는 한 번만 처리하는것으로 알고 있습니다.
이는 `Single`로 관리해도 좋지 않을까요?
서브젝트는 버튼 리스너와 같이 사용자 이벤트나 수시로 동작할 때 많이 사용됩니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+                super.onCreateSuccess(sessionDescription);
+                localPeer.setLocalDescription(new CustomSdpObserver("localSetLocal"), sessionDescription);
+                //SignalingClient.getInstance().emitSessionDescription(sessionDescription);
+                sdpSubject.onNext(sessionDescription);
```
Answer도 Offer와 같은 의견입니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/peer/manager/RendererManager.java:
```diff
@@ -0,0 +1,22 @@
+
+public class RendererManager {
+    private final EglBase rootEglBase;
+    public RendererManager(EglBase rootEglBase) {
```
singleton 객체를 보유하고있는 매니저들도 싱글톤으로 하면 어떨 것 같나요?
현재는 매니저가 매번 생성이 되서 초기화 하게되나요?

laco-dev commented on app/src/main/java/com/webrtc/boyj/api/peer/manager/VideoManager.java:
```diff
@@ -0,0 +1,61 @@
+    public VideoManager(Context context, EglBase rootEglBase, PeerConnectionFactory factory) {
+        this.context = context;
+/*        if (isCamera2Supported())
+            enumerator = new Camera2Enumerator(context);*/
```
카메라 2 서포트는 왜 주석처리가 되어있나요?

laco-dev commented on app/src/main/java/com/webrtc/boyj/utils/App.java:
```diff
@@ -0,0 +1,24 @@
+    private static Application sApplication;
+
+    private static Application getApplication() {
+        return sApplication;
```
밑에 Application Context를 반환하는게 있는데 Application 자체가 필요한 경우가 있나요?

laco-dev commented on app/src/main/java/com/webrtc/boyj/view/activity/CallActivity.java:
```diff
@@ -1,29 +1,57 @@
-        final User user = (User) getIntent().getSerializableExtra(Constants.EXTRA_USER);
+        Intent intent=getIntent();
+
+        boolean iscaller=intent.getBooleanExtra("caller",true);
```
EXTRA 값은 private static final로 해당 클래스에서 관리하는게 좋습니다.

jadenjack replied:
주로 카멜표기법을 사용한 것 같은데 isCaller가 더 어울리지 않을까요?

laco-dev commented on app/src/main/java/com/webrtc/boyj/view/activity/CallActivity.java:
```diff
@@ -1,29 +1,57 @@
+
         model.getIsActive().observe(this, isActive -> finish());
+        if(iscaller){
+            String deviceToken=intent.getStringExtra("deviceToken");
```
이러한 부분도 Key 값은 따로 상수로 보관하는게 좋아보입니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/view/activity/CallActivity.java:
```diff
@@ -1,29 +1,57 @@
+        rendererManager.initSurfaeView(binding.localView);
+        rendererManager.initSurfaeView(binding.remoteView);
+
         model.getIsActive().observe(this, isActive -> finish());
```
이 부분은 생각을 해봤는데
라이브 데이터는 데이터의 변화를 감지하면서 업데이트 하는 역할이 크다고 봅니다
라이브 데이터를 이용해 1회성 이벤트를 처리하는 방법이 있습니다.
이 점은 중요한 점은 아니지만 이후 고민을 좀 더 해보겠습니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/view/activity/CallActivity.java:
```diff
@@ -1,29 +1,57 @@
+        }
+
+
+        PeerConnectionClient.getInstance().remoteMediaStreamSubject.subscribe(mediaStream -> {
```
미디어 스트림은 어떤 스레드에서 동작하나요?
스케줄러를 통해 구분이 필요해 보입니다.

laco-dev commented on app/src/main/res/layout/activity_call.xml:
```diff
@@ -15,6 +15,22 @@
+            android:id="@+id/local_view"
+            android:layout_width="match_parent"
+            android:layout_height="250dp"
+            android:layout_margin="20dp"
```
두 가지 서피스 뷰 렌더러는 스타일로 레이아웃 관리하면 좋을 듯 합니다.

bhw1994 replied:
정확한 의미를 모르겠습니다 .
구체적인 예시 부탁드립니다.

laco-dev commented on app/src/main/java/com/webrtc/boyj/view/activity/CallActivity.java:
```diff
@@ -1,29 +1,57 @@
+            Logger.d("remotemediastream received");
+            VideoTrack videoTrack = mediaStream.videoTracks.get(0);
+            videoTrack.addSink(binding.remoteView);
+            PeerConnectionClient.getInstance().localStreamTo(binding.localView);
```
CallActivity가 PeerConnection Client에게 결국 로컬뷰를 넘겨주는데  이 작업은
Client에서 처리하도록 해도 되지 않을까요? 이럴거면 왜 구독하고 있는지 모르겠습니다.

bhw1994 replied:
PeerConnection Client 에서 처리하려면 멤버변수로 뷰를 가지고 있어야 합니다.
PeerConnection Client 가 뷰를 가지고 있는것은 관점이 섞인다고 판단하여 액티비티에서 구독하는 방식이 낫다고 판단하였습니다.

laco-dev reviewed (commented):
1. 모든 기능이 완성된 이후 필요 없는 부분을 억지로 주석처리하여 구분한 것으로 보입니다.
주석으로 모두 살아있어서 리뷰에 혼란이 옵니다.  다음에는 아예 빼고 올려주시거나 중간중간 작업 진행 부탁드립니다.
2.  RxJava에 대해 모든 dispose가 보이지 않습니다. 라이프사이클과 스케줄러 관리가 필요합니다.

jadenjack commented on app/src/main/java/com/webrtc/boyj/utils/Util.java:
```diff
@@ -0,0 +1,38 @@
+import retrofit2.Retrofit;
+import retrofit2.converter.gson.GsonConverterFactory;
+
+public class Util {
```
Util 클래스라는 자체 이름은 너무 일반적이라고 생각합니다. 조금 더 구체화된 클래스명으로 수정하는건 어떤가요

seokhyun-oh replied:
retrofit을 static 필드에서 초기화 한다면 Util 클래스가 싱글톤일 필요는 없어 보입니다.

jadenjack reviewed (commented):
PR을 설명하는 곳에 
**issue #2
이렇게 써있던데 혹시 헤더를 나타내는 # 문법과 헷갈린게 아닐까라는 생각에 첨언합니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/data/stun/IceServer.java:
```diff
@@ -0,0 +1,27 @@
+                ", username='" + username + '\'' +
+                ", credential='" + credential + '\'' +
+                '}';
+    }
```
```String.format()```을 고려해보는건 어떨까요?

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/data/IceServer.java:
```diff
@@ -0,0 +1,13 @@
+    }
+    public String toString() {
+        return uri + "[" + hostname + ":" + "]";
+    }
```
```String.format()```을 고려하는건 어떨까요?

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/data/stun/TurnServerPojo.java:
```diff
@@ -0,0 +1,30 @@
+    public String p;
+    @SerializedName("e")
+    @Expose
+    public Object e;
```
프로퍼티 명에 대한 추가적인 설명이 있었으면 좋겠습니다.
최소한 각 프로퍼티 모두에 대한 설명이 아니어도 어디서 참고하였다 정도는 필요합니다.

또한, 해당 프로퍼티들은  [SD](https://tools.ietf.org/html/rfc4566#section-5)(Session Description)으로부터 만들어진 것으로 추측되는데
그럴 경우 클래스 명이 적절하지 않다고 생각합니다.

seokhyun-oh replied:
추가적으로 해당 스팩을 볼 경우 각 프로퍼티의 종류와 의미,
필수로 갖고 있어야 할 정보와 선택적으로 가질 수 있는 정보,
하나의 값인지 여러 값을 가질 수 있는지를 명확히 파악할 수 있습니다.

이를 기반으로 Builder를 만들면 훨씬 재사용성이 높아질거같은데 어떻게 생각하시나요?

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/CustomPeerConnectionObserver.java:
```diff
@@ -0,0 +1,96 @@
+    public void onSignalingChange(PeerConnection.SignalingState signalingState) {
+        Logger.d("CustomPeerConnectionObserver : CustomPeerConnectionObserver");
+        //Log.d(logTag, "onSignalingChange() called with: signalingState = [" + signalingState + "]");
+    }
```
불필요한 주석은 제거해도 좋을거 같습니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/CustomPeerConnectionObserver.java:
```diff
@@ -0,0 +1,96 @@
+import org.webrtc.RtpTransceiver;
+
+public class CustomPeerConnectionObserver implements PeerConnection.Observer {
+    private String logTag = this.getClass().getCanonicalName();
```
불필요한 초기화 구문인것 같습니다.
해당 부분을 생성자에 넣으면 어떨까요?

```java
public CustomPeerConnectionObserver(final String logTag) {
  final String canonicalName = this.getClass().getCanonicalName();
  this.logTag = String.format("%s %s", canonicalName, logTag);
}
```

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/CustomSdpObserver.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public void onCreateSuccess(SessionDescription sessionDescription) {
+        Log.d(tag, "onCreateSuccess() called with: sessionDescription = [" + sessionDescription + "]");
```
로깅 클래스는 통일하면 좋을거같습니다.
[Logger](https://github.com/the-boyj/WebRTC_Android/blob/develop/app/src/main/java/com/webrtc/boyj/utils/Logger.java#L26)의 경우 내부적으로 stackTrace를 이용하기 때문에 추가적인 정보를 넘길 필요가 없습니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+import retrofit2.Response;
+
+public class PeerConnectionClient {
+    private static PeerConnectionClient instance;
```
Singleton 패턴을 염두한다면 static field에 저장할 필요는 없어보입니다.

seokhyun-oh replied:
오개념이 있었네요. 지우겠습니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+
+    private PeerConnectionClient(){
+        init();
+    }
```
```init()``` 메소드의 내용을 생성자에 넣거나
static 변수를 사용한다면 static 필드에서 초기화가 가능하므로 필요 없는 메소드라고 생각합니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+                    instance = new PeerConnectionClient();
+            }
+        return instance;
+    }
```
static 변수와 싱글톤 패턴 중 하나를 선택해서 구현하면 좋을거같습니다.

seokhyun-oh replied:
오개념이었습니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+        frontCapturer = videoManager.getFrontCameraCapturer();
+        backCapturer = videoManager.getBackCameraCapturer();
+        videoTrack = videoManager.getVideoTrack();
+        return true;
```
video에 대한 책임이 videoManager에 있는것 같습니다.
따라서 PeerConnectionClient가 frontCapturer등의 videoTrack 정보를 갖고있을 필요는 없어보입니다.
(필요시 호출하면 될거같네요)

seokhyun-oh replied:
boolean type을 반환하는 것이 무슨 역할을 하는지 모르겠습니다.
설명해주세요.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+
+        constraints = new MediaConstraints();
+        constraints.mandatory.add(new MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"));
+        constraints.mandatory.add(new MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"));
```
이 메소드는 생성자 혹은 static 필드로 대체 가능한것으로 보입니다.

seokhyun-oh replied:
설령 해당 메소드를 사용한다 해도 public 접근제어자는 부적절해보입니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+            }
+        });
+        addStreamToLocalPeer();
+    }
```
```createPeerConnection()```을 어디서 사용하는지 알 수 있을까요?
호출하는곳을 찾을 수 없네요.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+                localPeer.setRemoteDescription(new CustomSdpObserver("localSetRemote"), new SessionDescription(SessionDescription.Type.OFFER, data.getString("sdp")));
+            else
+                localPeer.setRemoteDescription(new CustomSdpObserver("localSetLocal"), new SessionDescription(SessionDescription.Type.ANSWER, data.getString("sdp")));
+        } catch(Exception e) { e.printStackTrace(); }
```
가급적 한줄짜리 제어문에도 중괄호와 개행을 사용하는것이 여러 이점이 있습니다.
```java
if (isOffer) {
  localPeer.setRemoteDescription(new ...);
}
```

seokhyun-oh replied:
```e.printStackTrace()```는 에러 핸들링이 아닙니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+    public void addIceCandidate(JSONObject data) {
+        try {
+            localPeer.addIceCandidate(new IceCandidate(data.getString("id"), data.getInt("label"), data.getString("candidate")));
+        } catch (JSONException e) { e.printStackTrace(); }
```
```e.printStackTrace()```는 예외 핸들링이 아닙니다.
만약 data 객체가 id, label, candidate중 하나를 포함하고 있지 않으면 아무 경고 없이 메소드가 종료될겁니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+        try { data = ("tbtzpdlql:0e7392fa-07ef-11e9-a443-0242ac110003").getBytes("UTF-8"); }
+        catch (UnsupportedEncodingException e) { e.printStackTrace(); }
+        String authToken = "Basic " + Base64.encodeToString(data, Base64.NO_WRAP);
+        Util.getInstance().getRetrofit().getIceCandidates(authToken).enqueue(new Callback<TurnServerPojo>() {
```
try{} 구문에서 에러가 발생할 경우 어떻게 작동하는지 설명 부탁드립니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+            public void onResponse(@NonNull Call<TurnServerPojo> call, @NonNull Response<TurnServerPojo> response) {
+                TurnServerPojo body = response.body();
+                if (body != null) {
+                    for (IceServer iceServer : body.iceServerList.iceServers) {
```
[링크](https://github.com/the-boyj/WebRTC_Android/pull/7/files#diff-6e6f6198ec7095477af1668efd5aacc6R26)에서 TurnServerPojo.iceServerList.iceServers가 null로 초기화 되는데
별도의 초기화 구문이 있는건가요?

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/PeerConnectionClient.java:
```diff
@@ -0,0 +1,262 @@
+            }
+            @Override
+            public void onFailure(@NonNull Call<TurnServerPojo> call, @NonNull Throwable t) {
+                t.printStackTrace();
```
해당 구문은 에러 핸들링이 가능한 구문이 아닙니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/manager/EglBaseManager.java:
```diff
@@ -0,0 +1,20 @@
+            }
+        }
+        return eglBase;
+    }
```
static final EglBase를 static 필드에서 초기화 한다면 매번 조건문을 거칠 필요가 없어 보입니다.
또한 아래 구문에서는 eglBase가 한번만 초기화 되므로 객체 생성에 대한 동기화 이슈를 고려하지 않아도 됩니다.

```java
private static final Eglbase eglBase;

static {
  eglBase = EglBase.create();
}

public static EglBase getEglBase() {
  return eglBase;
}
```

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/api/peer/manager/VideoManager.java:
```diff
@@ -0,0 +1,61 @@
+                return videoCapturer;
+            }
+        }
+        return null;
```
null이 아닌 예외와 try-catch 구문을 적절히 활용하면 Front, Back camera중 하나만 초기화 할 수 있을거같습니다.

seokhyun-oh commented on app/src/main/java/com/webrtc/boyj/view/activity/MainActivity.java:
```diff
@@ -12,13 +18,21 @@
     private MainAdapter adapter;
 
+    private final static String CAMERA = Manifest.permission.CAMERA;
+    private final static String RECORD_AUDIO = Manifest.permission.RECORD_AUDIO;
```
[Static Import](https://docs.oracle.com/javase/7/docs/technotes/guides/language/static-import.html)를 사용한다면 이미 상수로 선언 된 값들을 별도의 상수로 초기화 할 필요가 없어집니다.
